freegs/jtor.py

```diff
@@ -136,36 +137,57 @@ def fvac(self) -> float:
         pass
 
 
-class ConstrainBetapIp(Profile):
+class ProfileShapeFunction(abc.ABC):
+    """When instantiated, can be called with the normalised psi to
+    shape the profiles.
     """
-    Constrain poloidal Beta and plasma current
 
-    This is the constraint used in
-    YoungMu Jeon arXiv:1503.03135
+    @abc.abstractmethod
+    def __call__(self, psi_norm):
+        """
+        psi_norm - normalised psi [0, 1]
+        """
+        pass
+
+
+class DoublePowerShapeFunction(ProfileShapeFunction):
+    def __init__(self, alpha_m=1.0, alpha_n=2.0) -> None:
+        # Check inputs
+        if alpha_m < 0:
+            raise ValueError("alpha_m must be positive")
+        if alpha_n < 0:
+            raise ValueError("alpha_n must be positive")
 
+        self.alpha_m = alpha_m
+        self.alpha_n = alpha_n
+
+    def __call__(self, psi_norm):
+        return (1.0 - psi_norm**self.alpha_m) ** self.alpha_n
+
+
+class ConstrainBetapIpArbShape(Profile):
+    """Constrain poloidal Beta and plasma current with an arbitrary
+    shaping function
     """
 
-    def __init__(self, eq, betap, Ip, fvac, alpha_m=1.0, alpha_n=2.0, Raxis=1.0):
+    def __init__(self, eq, betap, Ip, fvac, shape_function=None, Raxis=1.0):
         """
         betap - Poloidal beta
         Ip    - Plasma current [Amps]
         fvac  - Vacuum f = R*Bt
-
+        shape_function - a callable which accepts a scalar normalised psi and array of alpha.
+        Defaults to `DoublePowerShapeFunction`.
```

Comment:
It just takes `psi`, right?

```suggestion
        shape_function - a callable which accepts a scalar normalised psi.
                         Defaults to `DoublePowerShapeFunction`.
```